test/src/tests-session-manager.ts

```diff
@@ -0,0 +1,860 @@
+            // FIXME: Test only passes if specific cases are set
+            it.skip('should log an AbandonedEndSession message if devToken is not set', () => {
+                mParticle.init(apiKey, window.mParticle.config);
+
+                const mpInstance = mParticle.getInstance();
+                sinon
+                    .stub(mpInstance._Persistence, 'getPersistence')
+                    .returns({ gs: {} });
+
+                const consoleSpy = sinon.spy(mpInstance.Logger, 'verbose');
+
+                // FIXME: Test only passes if devToken is undefined and webviewBridgeEnabled is false
+                mpInstance._Store.devToken = undefined;
+                mpInstance._Store.webviewBridgeEnabled = false;
+
+                mpInstance._SessionManager.endSession();
+
+                // Should log initial StartingEndSession and AbandonEndSession messagesk
+                expect(consoleSpy.getCalls().length).to.equal(2);
+                expect(consoleSpy.lastCall.firstArg).to.equal(
+                    Messages.InformationMessages.AbandonEndSession
+                );
+            });
+
+            it('should priortize cookie sessionId over Store.sessionId', () => {
+                mParticle.init(apiKey, window.mParticle.config);
+
+                const mpInstance = mParticle.getInstance();
+                sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                    gs: {
+                        sid: 'cookie-session-id',
+                    },
+                });
+
+                mpInstance._Store.sessionId = 'store-session-id';
+
+                mpInstance._SessionManager.endSession();
+
+                expect(mpInstance._Store.sessionId).to.equal(
+                    'cookie-session-id'
+                );
+            });
+
+            it('should NOT end session if session has not timed out', () => {
+                const now = new Date();
+                // The default timeout limit is 30 minutes.
+                const twentyMinutesAgo = new Date();
+                twentyMinutesAgo.setMinutes(now.getMinutes() - 20);
+
+                mParticle.init(apiKey, window.mParticle.config);
+                const mpInstance = mParticle.getInstance();
+                const timerSpy = sinon.spy(
+                    mpInstance._SessionManager,
+                    'setSessionTimer'
+                );
+
+                // Session Manager relies on persistence to determine last time seen (LES)
+                // Also requires sid to verify session exists
+                sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                    gs: {
+                        les: twentyMinutesAgo,
+                        sid: 'fake-session-id',
+                    },
+                });
+
+                mpInstance._SessionManager.endSession();
+
+                expect(mpInstance._Store.sessionId).to.equal('fake-session-id');
```

Comment:
I was a little thrown off by this, but I realize that if the session ended, the `sessionId` would be `null`.  Maybe add that as a comment as to what you are testing here, or add an `expect` right before this line to say that the `sessionId` doesn't equal `null`.